iis-wf/iis-wf-report/src/main/java/eu/dnetlib/iis/wf/report/pushgateway/process/PushMetricsProcess.java

```diff
@@ -0,0 +1,331 @@
+
+    private MetricPusherCreatorProducer metricPusherCreatorProducer = new MetricPusherCreatorProducer();
+    private MetricPusherProducer metricPusherProducer = new MetricPusherProducer();
+    private FileSystemProducer fileSystemProducer = new FileSystemProducer();
+    private ReportLocationsFinder reportLocationsFinder = new ReportLocationsFinder();
+    private LabelNamesByMetricNameProducer labelNamesByMetricNameProducer = new LabelNamesByMetricNameProducer();
+    private ReportEntryReader reportEntryReader = new ReportEntryReader();
+    private ReportEntryConverter reportEntryConverter = new ReportEntryConverter();
+    private GaugesRegistrar gaugesRegistrar = new GaugesRegistrar();
+    private JobNameProducer jobNameProducer = new JobNameProducer();
+
+    @Override
+    public Map<String, PortType> getInputPorts() {
+        return Collections.emptyMap();
+    }
+
+    @Override
+    public Map<String, PortType> getOutputPorts() {
+        return Collections.emptyMap();
+    }
+
+    @Override
+    public void run(PortBindings portBindings, Configuration conf, Map<String, String> parameters) {
+        metricPusherCreatorProducer.create(parameters)
+                .ifPresent(metricPusherCreator -> {
+                    logger.info("Using MetricPusherCreatorProducer: {}", metricPusherCreator);
+                    metricPusherProducer.create(metricPusherCreator, parameters)
+                            .ifPresent(metricPusher -> {
+                                logger.info("Using MetricPusher: {}", metricPusher);
+                                fileSystemProducer.create(conf)
+                                        .ifPresent(fs -> {
+                                            logger.info("Using FileSystem: {}", fs);
+                                            reportLocationsFinder.find(parameters)
+                                                    .ifPresent(paths -> {
+                                                        logger.info("Using report locations of size: {}", paths.size());
+                                                        labelNamesByMetricNameProducer.create(parameters)
+                                                                .ifPresent(labelNamesByMetricName -> {
+                                                                    logger.info("Using label map of size: {}", labelNamesByMetricName.size());
+                                                                    List<Optional<List<Gauge>>> optionals = paths.stream()
+                                                                            .map(path -> reportEntryReader.read(fs, new Path(path))
+                                                                                    .flatMap(reportEntries -> reportEntryConverter.convert(reportEntries, path, labelNamesByMetricName)))
+                                                                            .collect(Collectors.toList());
```

Comment:
Note: This part of code is refactored but the comment is still valid. It would seem like a good solution but I think the better solution is to stay with `Optional`. In our new approach we specify as input for the reader a dir containing reports as subdirs. The dir is the same for all workflows, including the one that generates summary json report. But reading json report using avro reader fails. We can cleanly support this situation with non-empty optional for successful reading and empty optional for unsuccessful reading.